### PR TITLE
NIFI-9657 Create MoveADLS processor

### PR DESCRIPTION
**Pull request by timeabarna**
https://issues.apache.org/jira/browse/NIFI-9657

#### Description of PR

Enables moving content into an Azure Data Lake Storage Gen 2.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [x] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by Lehel44** [comment]
Reviewing

**Review by turcsanyip — CHANGES_REQUESTED** [review]
@timeabarna Thanks for adding this new feature. I haven't built / tested the processor, just spotted some improvement point in the code. Please find them below in the inline comments.

**Comment by timeabarna** [comment]
Thanks @turcsanyip for your help, I've modified the code based on your recommendations.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/MoveAzureDataLakeStorage.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+                final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+                session.getProvenanceReporter().send(flowFile, fileClient.getFileUrl(), transferMillis);
+            } catch (DataLakeStorageException dlsException) {
+                if (dlsException.getStatusCode() == 409) {
+                    if (conflictResolution.equals(IGNORE_RESOLUTION)) {
+                        session.transfer(flowFile, REL_SUCCESS);
+                        String warningMessage = String.format("File with the same name already exists. " +
+                                "Remote file not modified. " +
+                                "Transferring {} to success due to %s being set to '%s'.", CONFLICT_RESOLUTION.getDisplayName(), conflictResolution);
+                        getLogger().warn(warningMessage, new Object[]{flowFile});
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/MoveAzureDataLakeStorage.java:209]: You can remove the array creation here and just simply pass the args.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/MoveAzureDataLakeStorage.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+                attributes.put(ATTR_NAME_FILENAME, fileName);
+                attributes.put(ATTR_NAME_PRIMARY_URI, fileClient.getFileUrl());
+                attributes.put(ATTR_NAME_LENGTH, String.valueOf(fileClient.getProperties().getFileSize()));
+                flowFile = session.putAllAttributes(flowFile, attributes);
+
+                session.transfer(flowFile, REL_SUCCESS);
+                final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+                session.getProvenanceReporter().send(flowFile, fileClient.getFileUrl(), transferMillis);
+            } catch (DataLakeStorageException dlsException) {
+                if (dlsException.getStatusCode() == 409) {
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/MoveAzureDataLakeStorage.java:203]: Do you think the two condition branches can be unified like this?

`if (dlsException.getStatusCode() == 409 && conflictResolution.equals(IGNORE_RESOLUTION) {
...
} else throw dlsException`

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/MoveAzureDataLakeStorage.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            final DataLakeFileSystemClient sourceFileSystemClient = storageClient.getFileSystemClient(sourceFileSystem);
+            final DataLakeDirectoryClient sourceDataLakeDirectoryClient = sourceFileSystemClient.getDirectoryClient(sourceDirectory);
+            final DataLakeFileSystemClient destinationFileSystemClient = storageClient.getFileSystemClient(destinationFileSystem);
+            final DataLakeDirectoryClient destinationDirectoryClient = destinationFileSystemClient.getDirectoryClient(destinationDirectory);
+            DataLakeFileClient fileClient = sourceDataLakeDirectoryClient.getFileClient(fileName);
+            final DataLakeRequestConditions sourceConditions = new DataLakeRequestConditions();
+            final DataLakeRequestConditions destConditions = new DataLakeRequestConditions();
+            final String conflictResolution = context.getProperty(CONFLICT_RESOLUTION).getValue();
+
+            try {
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/MoveAzureDataLakeStorage.java:172]: I think this nested try may be extracted to a separate method to make the code more readable.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/MoveAzureDataLakeStorage.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        final long startNanos = System.nanoTime();
+        try {
+            final String sourceFileSystem = evaluateFileSystemProperty(context, flowFile, SOURCE_FILESYSTEM);
+            final String sourceDirectory = evaluateDirectoryProperty(context, flowFile, SOURCE_DIRECTORY);
+            final String destinationFileSystem = evaluateFileSystemProperty(context, flowFile, DESTINATION_FILESYSTEM);
+            final String destinationDirectory = evaluateDirectoryProperty(context, flowFile, DESTINATION_DIRECTORY);
+            final String fileName = evaluateFileNameProperty(context, flowFile);
+
+            String destinationPath = destinationDirectory;
+            if (!destinationDirectory.isEmpty() && ! sourceDirectory.isEmpty()) {
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/MoveAzureDataLakeStorage.java:158]: Would you please code format this class?

**Comment by timeabarna** [comment]
Thanks @Lehel44 for your review. I've modified the code based on our discussion.

**Comment by Lehel44** [comment]
Thanks @timeabarna LGTM+1

**Review by turcsanyip — CHANGES_REQUESTED** [review]
@timeabarna I found some more improvement points. Could you please check them?

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java [hunk 5]
```diff
@@ -200,19 +208,30 @@ public static String evaluateFileNameProperty(ProcessContext context, FlowFile f
...
+             this.displayName = null;
+         }
+
+         public DirectoryValidator(String displayName) {
+             this.displayName = displayName;
+         }
+
+         @Override
         public ValidationResult validate(String subject, String input, ValidationContext context) {
+            displayName = displayName == null ? DIRECTORY.getDisplayName() : displayName;
```
[the remark this answers is not in the document]
Comment (timeabarna) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureDataLakeStorageProcessor.java:224]: Thanks @turcsanyip for your help. I had the same approach first then discovered that sadly DIRECTORY property was not build yet at the constructor time so it is not possible to use as a value at this point.

**Comment by timeabarna** [comment]
@turcsanyip Thanks for your review, code has been modified based on your findings except the DirectoryValidator.

**Pull request closed without merging**